### PR TITLE
Integrate emilio

### PR DESCRIPTION
**Pull request by iilyak**
## Overview

Add `emilio` erlang linter to ensure unified style.

## Testing recommendations

- do a fresh clone and run `./configure` (make sure `bin/emilio` script is created)
- run `./configure` again and make sure it doesn't try to install `emilio` again
- Update any erlang files in any dependency (except files mentioned in `ignore` section of `emilio.config`)
- Run `make emilio` and make sure that:
  - produced warnings are only for (± 3) updated lines
  - make fails with non 0 exit code  
- Run `bin/emilio -c emilio.config src | bin/warnings_in_scope -s 0` to see warnings from `emilio`
  - make sure produced warnings are only for updated lines
- Run `bin/emilio -c emilio.config src` and make sure dependencies mentioned in `ignore` section of `emilio.config` are ignored
- test it on windows if you have access to it (I don't have access to windows :-()  

## Checklist

- [x] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -217,12 +217,25 @@ install_local_rebar() {
     fi
 }
 
+install_local_emilio() {
+    if [ ! -x "${rootdir}/bin/emilio" ]; then
+        if [ ! -d "${rootdir}/src/emilio" ]; then
```
Comment (wohali) [inline configure]: Probably need to add `src/emilio` and `bin/emilio` to `.gitgnore` with this approach.

**Comment by wohali** [comment]
Interesting, will look at this in more detail later

**Comment by iilyak** [comment]
@wohali I know you are busy with 3.0. Should I rebase against latest master or wait couple months (the release process is exhausting I am sure)?

**Comment by wohali** [comment]
Hey @iilyak , it was @davisp who was initially against this. If you can get his +1, I'd love to see it in 3.0, go ahead and rebase.

File: Makefile.win [hunk 2]
```diff
@@ -124,9 +124,10 @@ fauxton: share\www
 check: all
 	@$(MAKE) -f Makefile.win test-cluster-with-quorum
 	@$(MAKE) -f Makefile.win test-cluster-without-quorum
+	@$(MAKE) -f Makefile.win emilio
+	@$(MAKE) -f Makefile.win python-black
```
Comment (wohali) [inline Makefile.win]: Can you make this match the order of `Makefile`? Thanks.

File: .gitignore [hunk 3]
```diff
@@ -44,6 +44,7 @@ src/couch/priv/couch_js/util.d
 src/couch/priv/icu_driver/couch_icu_driver.d
 src/mango/src/mango_cursor_text.nocompile
 src/docs/
+src/emilio/
```
Comment (wohali) [inline .gitignore]: Add `bin/emilio` too

File: configure [hunk 4]
```diff
@@ -217,12 +217,25 @@ install_local_rebar() {
     fi
 }
 
+install_local_emilio() {
```
Comment (wohali) [inline configure]: You need to add all of this to `configure.ps1` too. Are you up to the challenge? 🥊

File: .gitignore (hunk 3 above)
Comment (iilyak) [reply .gitignore]: bin/ directory should be ignored already

File: configure (hunk 4 above)
Comment (iilyak) [reply configure]: This is a challenge indeed.

**Comment by garrensmith** [comment]
I'm digging this up again. Could we look at adding this into the CI flow without adding support for windows? Its going to be hard to do that.

**Comment by wohali** [comment]
@garrensmith If you or @iilyak can freshen this PR to not have a conflict on `Makefile`, I'll have a look at `configure.ps1`.

**Comment by wohali** [comment]
Also, if you can move the source branch into the `apache/couchdb` repo, I can PR against it with the Windows changes. Otherwise, I'll just post a diff here that you can apply yourself.

**Comment by garrensmith** [comment]
Awesome thanks that will be great

________________________________

File: configure [hunk 5]
```diff
@@ -217,12 +217,25 @@ install_local_rebar() {
...
+    if [ ! -x "${rootdir}/bin/emilio" ]; then
+        if [ ! -d "${rootdir}/src/emilio" ]; then
+            git clone --depth 1 https://github.com/cloudant-labs/emilio ${rootdir}/src/emilio
+        fi
+        cwd=`pwd`
+        cd ${rootdir}/src/emilio && ${REBAR} compile escriptize; cd ${cwd}
+        mv ${rootdir}/src/emilio/emilio ${rootdir}/bin/emilio
+        chmod +x ${rootdir}/bin/emilio
```
Comment (wohali) [inline configure:228]: Should also do a `cd ${rootdir}/src/emilio && ${REBAR} clean` at this point, like we do on L216 for `rebar`.

File: configure (hunk 4 above)
Comment (wohali) [reply configure:220]: Here's the diff, feel free to add to this PR:

```diff
diff --git a/configure.ps1 b/configure.ps1
index c74fbcf41..21c079f67 100644
--- a/configure.ps1
+++ b/configure.ps1
@@ -205,6 +205,20 @@ if ((Get-Command "rebar.cmd" -ErrorAction SilentlyContinue) -eq $null)
    $env:Path += ";$rootdir\bin"
 }

+# check for emilio; if not found, get it, build it, add it to path
+if ((Get-Command "emilio.cmd" -ErrorAction SilentlyContinue) -eq $null)
+{
+   Write-Verbose "==> emilio.cmd not found; bootstrapping..."
+   if (-Not (Test-Path "src\emilio"))
+   {
+      git clone --depth 1 https://github.com/wohali/emilio $rootdir\src\emilio
+   }
+   cmd /c "cd $rootdir\src\emilio && rebar compile escriptize"
+   cp $rootdir\src\emilio\emilio $rootdir\bin\emilio
+   cp $rootdir\src\emilio\bin\emilio.cmd $rootdir\bin\emilio.cmd
+   cmd /c "cd $rootdir\src\emilio && rebar clean"
+}
+
 # only update dependencies, when we are not in a release tarball
 if ( (Test-Path .git -PathType Container) -and (-not $SkipDeps) ) {
     Write-Verbose "==> updating dependencies"
```

Note this depends on https://github.com/cloudant-labs/emilio/pull/3 being merged.

**Comment by davisp** [comment]
@wohali cloudant-labs/emilio#3 is merged.

File: configure (hunk 4 above)
Comment (iilyak) [reply configure:220]: Thank you very much.

**Comment by iilyak** [comment]
CI is failing with unrelated issue 😢 
```
[2020-04-03T11:40:38.315Z] ==> couch_replicator (compile)

[2020-04-03T11:40:38.662Z] /home/****/workspace/****-cm1_PullRequests_PR-1944/src/couch_replicator/src/couch_replicator_httpc.erl:16: can't find include lib "ibrowse/include/ibrowse.hrl"
```

**Comment by wohali** [comment]
@iilyak 

I also can't build your branch, either in the CI environment or locally. Here's what I'm doing:

```
rm -rf couchdb
git clone https://github.com/cloudant/couchdb
cd couchdb
git checkout integrate-emilio
./configure --with-curl
make dist
```

This fails in the same way as Jenkins is failing.

I don't get the same failure from `apache/couchdb:master`.

Also you still need to apply the diff I provided in comment.

Perhaps close this PR and reopen it by doing the branch off of `apache/couchdb` instead of `cloudant/couchdb`? Maybe it's because `cloudant/couchdb:master` is ~1400 commits behind, I don't know. @rnewson might have a better idea.

**Comment by jaydoane** [comment]
Possibly of interest: https://medium.com/@elbrujohalcon/are-formatters-better-than-linters-cbab91189be3

**Comment by iilyak** [comment]
@wohali: Thank you for your help with Windows part.

**Pull request merged**